### DISCUSSION TITLE
Accessing $id in a component's init function

### DISCUSSION THREAD

**Discussion by docdunning** [alpinejs/alpine#4171] [1. Help]
I'm using `$id` to generate unique ids, and it's working fine in that respect.

However, I'd like to access the id when the component is initialised, and that doesn't quite work as I expected.

See the example code below.  In the component's `init()` function, it logs `this.$el` to the console, which shows that the id has been generated:

`<div x-data="block" :id="$id('block')" id="block-1">Block 1</div>`

Then it logs `this.$el.id`, which I expect to show `block-1`, but it's an empty string.  

As per the code below, I can use `$nextTick` to log `this.$el.id`, in which case it does show the correct value for the id.

Is `$nextTick` the best solution, or have I misunderstood something about the sequence of events?

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.10/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="block" :id="$id('block')">Block 1</div>
<div x-data="block" :id="$id('block')">Block 2</div>
</body>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.data('block',() => ({
    init() {
      console.log(this.$el, this.$el.id)
      this.$nextTick(() => console.log(this.$el.id))
    }
  }))
})
</script>
</html> 
```

**Comment by SimoTod** [comment]
Yes, you are correct. init runs before all the other directives are computed (except for x-data). It's kind of the equivalent to the old beforeMount in VueJs/React. If you want to do it after (aka component mounted), nextTick is the right solution.

AN alternative solution is to use id in x-data

**Reply by docdunning** [reply]
@SimoTod 

Thanks for confirming.

Could you explain a little more what you mean by "use id in x-data" as an alternative?

**Reply by SimoTod** [reply]
As shown in the first example [here](https://alpinejs.dev/magics/id#groups-with-x-id) but I'm not sure it would play nicely with Alpine.data. I'd stick to nextTick

**Reply by docdunning** [reply]
>  I'd stick to nextTick

Understood - thanks :-)
